tests/git/utils.ts

```diff
@@ -1,72 +1,270 @@
+        Buffer.from(data.objectId),
+        gitUtils.LINE_FEED_BUFFER,
+      ]);
+      return Buffer.concat([gitHttp.packetLineBuffer(line), rest]);
+    }
+    case 'SEPARATOR':
+      return Buffer.concat([Buffer.from('0000'), rest]);
+    case 'done':
+      return Buffer.concat([Buffer.from('0009done\n'), rest]);
+    case 'none':
+      return Buffer.alloc(0);
+    default:
+      utils.never();
+  }
+}
+
+// Used to print out the contents of an `Buffer` iterable for testing
+async function* tapGen(
+  gen: AsyncIterable<Buffer>,
+): AsyncGenerator<Buffer, void, void> {
+  let acc = '';
+  for await (const line of gen) {
+    acc += line.toString();
+    yield line;
+  }
+  // eslint-disable-next-line no-console
+  console.log(acc);
+}
+
+/**
+ * Create a test request handler for use with `git.clone` and `git.pull`
+ */
+function request({
+  fs,
+  dir,
+  gitDir,
+}: {
+  fs: any;
+  dir: string;
+  gitDir: string;
+}) {
+  return async ({
+    url,
+    method = 'GET',
+    headers = {},
+    body = [Buffer.from('')],
+  }: {
+    url: string;
+    method: string;
+    headers: POJO;
+    body: Array<Buffer>;
+  }) => {
+    // Console.log('body', body.map(v => v.toString()))
+    if (method === 'GET') {
+      // Send back the GET request info response
+      const advertiseRefGen = gitHttp.advertiseRefGenerator({
+        fs,
+        dir,
+        gitDir,
+      });
+
+      return {
+        url: url,
+        method: method,
+        body: advertiseRefGen,
+        headers: headers,
+        statusCode: 200,
+        statusMessage: 'OK',
+      };
+    } else if (method === 'POST') {
+      const packGen = gitHttp.generatePackRequest({
+        fs,
+        dir,
+        gitDir,
+        body,
+      });
+      return {
+        url: url,
+        method: method,
+        body: packGen,
+        headers: headers,
+        statusCode: 200,
+        statusMessage: 'OK',
+      };
+    } else {
+      utils.never();
+    }
+  };
 }
 
-export { createGitRepo, getPackID };
+const objectIdArb = fc.hexaString({
+  maxLength: 40,
+  minLength: 40,
+});
+const capabilityArb = fc.stringOf(
+  fc.constantFrom(...`abcdefghijklmnopqrstuvwxyz-1234567890`.split('')),
+  { minLength: 5 },
+);
+const capabilityListArb = fc.array(capabilityArb, { size: 'small' });
+const restArb = fc.uint8Array();
+const wantArb = fc.record({
+  type: fc.constant('want') as Arbitrary<'want'>,
+  objectId: objectIdArb,
+  capabilityList: capabilityListArb,
+});
+const haveArb = fc.record({
+  type: fc.constant('have') as Arbitrary<'have'>,
+  objectId: objectIdArb,
+});
+
+const lineDataArb = fc.oneof(
+  wantArb,
+  haveArb,
+  fc.record({
+    type: fc.constantFrom<'SEPARATOR' | 'done' | 'none'>(
+      'SEPARATOR',
+      'done',
+      'none',
+    ),
+  }),
+);
+
+export {
+  createGitRepo,
+  listGitObjects,
+  generateTestNegotiationLine,
+  tapGen,
+  request,
+  objectIdArb,
+  capabilityArb,
+  capabilityListArb,
+  restArb,
+  wantArb,
+  haveArb,
```

Comment:
I think all these names should be expanded a bit. Like `wantArb`? That's way too simple. It should be specific to particular data structures within git.

Reply:
I have cleaned these up. Added some comments and made the names more specific.